File: DSA/PriyankaSingh_590016827/day39question3.c

#include <stdio.h>

#define MAX 100

int arr[MAX];
int top1 = -1, top2 = MAX;

void push1(int x) {
    if (top1 + 1 == top2) {
        printf("Stack Overflow\n");
        return;
    }
    arr[++top1] = x;
}

void push2(int x) {
    if (top2 - 1 == top1) {
        printf("Stack Overflow\n");
        return;
    }
    arr[--top2] = x;
}

int pop1() {
    if (top1 == -1) return -1;
    return arr[top1--];
}

int pop2() {
    if (top2 == MAX) return -1;
    return arr[top2++];
}

void printStacks() {
    printf("Stack1: ");
    for (int i = 0; i <= top1; i++)
        printf("%d ", arr[i]);

    printf("\nStack2: ");
    for (int i = MAX - 1; i >= top2; i--)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    push1(1);
    push1(2);
    push2(9);
    push2(8);

    printStacks();

    push1(3);
    pop2();

    printStacks();

    return 0;
}

File: DSA/Pranvkumar_590011587/Day39_Question3.py
class TwoStacks:
    def __init__(self, size):
        self.size = size
        self.arr = [None] * size
        self.top1 = -1
        self.top2 = size
    def push1(self, x):
        if self.top1 < self.top2 - 1:
            self.top1 += 1
            self.arr[self.top1] = x
        else:
            print("Stack Overflow in Stack1")
    def push2(self, x):
        if self.top1 < self.top2 - 1:
            self.top2 -= 1
            self.arr[self.top2] = x
        else:
            print("Stack Overflow in Stack2")
    def pop1(self):
        if self.top1 >= 0:
            val = self.arr[self.top1]
            self.top1 -= 1
            return val
        else:
            print("Stack Underflow in Stack1")
    def pop2(self):
        if self.top2 < self.size:
            val = self.arr[self.top2]
            self.top2 += 1
            return val
        else:
            print("Stack Underflow in Stack2")
    def get_stack1(self):
        return self.arr[:self.top1+1]
    def get_stack2(self):
        return self.arr[self.top2:]
    
# Example Usage
ts = TwoStacks(10)
ts.push1(1)
ts.push1(2)
ts.push2(9)
ts.push2(8)
print("Stack1:", ts.get_stack1())  
print("Stack2:", ts.get_stack2()) 

ts.push1(3)
ts.pop2()
print("Stack1:", ts.get_stack1())  
print("Stack2:", ts.get_stack2())  